File: r/002.r
LIMIT <- 4*1000*1000

sum <- 0
a   <- 1
b   <- 2

while (b < LIMIT) {
    tmp <- a
    a   <- b
    b   <- tmp + b

    # On the last iteration, b will be greater than LIMIT and a will be the
    #  largest fib under LIMIT.  On the first iteration, a becomes 2 and gets
    #  added to sum.  If a's initial value were even, we would need to add it
    #  to sum before the loop.
    if (a %% 2 == 0) {
        sum <- sum + a
    }
}

print(sum)

File: go/002.go
package main

var nextFib = func() (func() int) {
	a, b := 0, 1

	return func() int {
		a, b = b, a + b
		return b
	}
}()

const MAX = 4 * 1000 * 1000

func main() {
	sum := 0

	for {
		f := nextFib()

		if f > MAX {
			break
		}

		if f % 2 == 0 {
			sum += f
		}
	}

	println(sum)
}